File: Java/Tema7_Ejercicio9.java
/*
 * TEMA 7: FUNCIONES Y MÉTODOS - EJERCICIO 9
 * ==========================================
 * Escriba un programa que use sobrecarga de funciones (overloading).
 */

public class Tema7_Ejercicio9 {
    
    // Sumar dos enteros
    static int sumar(int a, int b) {
        return a + b;
    }
    
    // Sumar tres enteros
    static int sumar(int a, int b, int c) {
        return a + b + c;
    }
    
    // Sumar dos doubles
    static double sumar(double a, double b) {
        return a + b;
    }
    
    // Sumar dos cadenas (concatenación)
    static String sumar(String a, String b) {
        return a + " " + b;
    }
    
    // Sumar arreglo
    static int sumar(int[] array) {
        int suma = 0;
        for (int num : array) {
            suma += num;
        }
        return suma;
    }
    
    public static void main(String[] args) {
        System.out.println("=== SOBRECARGA DE FUNCIONES (OVERLOADING) ===");
        
        System.out.println("\n--- Sumar enteros ---");
        System.out.println("sumar(5, 10) = " + sumar(5, 10));
        System.out.println("sumar(5, 10, 15) = " + sumar(5, 10, 15));
        
        System.out.println("\n--- Sumar doubles ---");
        System.out.println("sumar(5.5, 10.5) = " + sumar(5.5, 10.5));
        
        System.out.println("\n--- Concatenar cadenas ---");
        System.out.println("sumar(\"Hola\", \"Mundo\") = " + sumar("Hola", "Mundo"));
        
        System.out.println("\n--- Sumar arreglo ---");
        int[] numeros = {10, 20, 30, 40, 50};
        System.out.println("sumar({10, 20, 30, 40, 50}) = " + sumar(numeros));
    }
}

/*
 * SALIDA ESPERADA:
 * === SOBRECARGA DE FUNCIONES (OVERLOADING) ===
 * 
 * --- Sumar enteros ---
 * sumar(5, 10) = 15
 * sumar(5, 10, 15) = 30
 * 
 * --- Sumar doubles ---
 * sumar(5.5, 10.5) = 16.0
 * 
 * --- Concatenar cadenas ---
 * sumar("Hola", "Mundo") = Hola Mundo
 * 
 * --- Sumar arreglo ---
 * sumar({10, 20, 30, 40, 50}) = 150
 */

File: JS/Tema7_Ejercicio9.js
/*
 * TEMA 7: FUNCIONES Y MÉTODOS - EJERCICIO 9
 * =========================================
 * Sobrecarga de funciones.
 * 
 * EXPLICACIÓN:
 * - Simular sobrecarga usando parámetros opcionales
 * - Detectar tipos de argumento
 * - Comportamiento diferente según entrada
 */

function procesar(a, b) {
    // Suma de dos números
    if (typeof a === 'number' && typeof b === 'number') {
        return a + b;
    }
    
    // Concatenación de strings
    if (typeof a === 'string' && typeof b === 'string') {
        return a + " " + b;
    }
    
    // Repetir array
    if (Array.isArray(a) && typeof b === 'number') {
        return Array(b).fill(a).flat();
    }
    
    return "Tipos no soportados";
}

function info(valor) {
    const tipo = typeof valor;
    
    if (tipo === 'number') {
        return `Número: ${valor}`;
    } else if (tipo === 'string') {
        return `String: "${valor}"`;
    } else if (Array.isArray(valor)) {
        return `Array: [${valor}]`;
    } else {
        return `${tipo}: ${valor}`;
    }
}

console.log("=== SOBRECARGA DE FUNCIONES ===");

console.log("Función procesar():");
console.log(`procesar(10, 5) = ${procesar(10, 5)}`);
console.log(`procesar("Hola", "mundo") = "${procesar("Hola", "mundo")}"`);
console.log(`procesar([1, 2], 3) = [${procesar([1, 2], 3)}]`);

console.log("\nFunción info():");
console.log(info(42));
console.log(info("JavaScript"));
console.log(info([1, 2, 3]));

/*
 * SALIDA ESPERADA:
 * === SOBRECARGA DE FUNCIONES ===
 * Función procesar():
 * procesar(10, 5) = 15
 * procesar("Hola", "mundo") = "Hola mundo"
 * procesar([1, 2], 3) = [1,2,1,2,1,2]
 * 
 * Función info():
 * Número: 42
 * String: "JavaScript"
 * Array: [1,2,3]
 */
